spark3/src/main/java/org/apache/iceberg/spark/procedures/AddFilesProcedure.java

```diff
@@ -0,0 +1,319 @@
+    }
+
+    importPartitions(table, partitions);
+  }
+
+  private void importPartitions(Table table, List<SparkTableUtil.SparkPartition> partitions) {
+    String stagingLocation = table.properties()
+        .getOrDefault(TableProperties.WRITE_METADATA_LOCATION, table.location() + "/metadata");
+    SparkTableUtil.importSparkPartitions(spark(), partitions, table, table.spec(), stagingLocation);
+  }
+
+  private List<SparkTableUtil.SparkPartition> filterPartitions(List<SparkTableUtil.SparkPartition> partitions,
+                                                               Map<String, String> partition) {
+    if (partition.isEmpty()) {
+      // No partition filter arg
+
+      if (partitions.isEmpty()) {
+        throw new IllegalArgumentException("Cannot add files, no files found in the table.");
+      }
+      return partitions;
+    } else {
+      // Partition filter arg passed
+
+      List<SparkTableUtil.SparkPartition> filteredPartitions = partitions
+          .stream().filter(p -> p.getValues().equals(partition)) // Todo Support Wildcards
+          .collect(Collectors.toList());
+
+      if (filteredPartitions.isEmpty()) {
+        throw new IllegalArgumentException(
+            String.format("No partitions found in table for add_file command. " +
+                    "Looking for a partition with the value %s",
+                partition.entrySet().stream().map(e -> e.getKey() + "=" + e.getValue())
+                    .collect(Collectors.joining(",")))
+        );
+      }
+
+      return filteredPartitions;
+    }
+  }
+
+  @Override
+  public String description() {
+    return null;
+  }
+
+  private void validatePartitionSpec(Table table, Map<String, String> partition) {
+    List<PartitionField> partitionFields = table.spec().fields();
+
+    boolean tablePartitioned = !partitionFields.isEmpty();
+    boolean partitionSpecPassed = !partition.isEmpty();
+
+    if (tablePartitioned && partitionSpecPassed) {
+      // Has Partitions, Check that they are valid
+      if (partitionFields.size() != partition.size()) {
+        throw new IllegalArgumentException(
+          String.format(
+            "Cannot add data files to target table %s because that table is partitioned, " +
+            "but the number of columns in the provided partition spec (%d) " +
+            "does not match the number of partitioned columns in table (%d)",
+              table.name(), partition.size(), partitionFields.size()));
+      }
+      partitionFields.forEach(field -> {
+        if (!partition.containsKey(field.name())) {
+          throw new IllegalArgumentException(
+            String.format(
+              "Cannot add files to target table %s. %s is partitioned but the specified partition spec " +
+              "refers to a column that is not partitioned: '%s'",
+                table.name(), table.name(), field.name())
+            );
+        }
+      });
```

Comment:
I think we need to add validation that all columns from Iceberg partition spec are present in the partition map provided by the user to avoid cases mentioned by @rdblue.

```
add_files(
  table => 'db.iceberg_table', -- partitioned Iceberg table
  source_table => '`parquet`.`path/to/table/part_col_1=x`'
  partition => map('part_col_2', 'y')
)
```

Reply:
I'd say the `partition` map should be required for partitioned tables.

Reply:
Per discussion, this should act as a filter and can be optional.